### ISSUE TITLE
nix overlay support

### ISSUE THREAD

**Issue by caarlos0** [goreleaser/goreleaser#3537]
would be nice if goreleaser could create/manage a nix overlay... I've been told! 

I have no experience with nix, but this sounds like a cool idea given the little I understand about it...

**Comment by caarlos0** [comment]
cc/ @maaslalani

**Comment by a-h** [comment]
# Nixpkgs already has goreleaser

nixpkgs already has a setup for goreleaser, which you can run with nix using `nix shell nixpkgs#goreleaser` (modern flake-enabled CLI), or `nix-shell -p goreleaser` (older syntax, but still current).

The nix search 
https://search.nixos.org/packages?channel=22.11&show=goreleaser&from=0&size=50&sort=relevance&type=packages&query=goreleaser shows that 1.13 is the stable version.

The stable Nix configuration is at https://github.com/NixOS/nixpkgs/blob/nixos-22.11/pkgs/tools/misc/goreleaser/default.nix while the unstable version is bang up to date at v1.16.0 - https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/misc/goreleaser/default.nix

# Issues with nixpkgs

The goreleaser Nix maintainers are doing a great job (thanks!) of keeping up to date, but having a central nixpkgs repo has some drawbacks in terms of getting releases merged into the central repo, and also, I often want to update just one tool (like `goreleaser`), not take in all of the downstream updates.

With nixpkgs, I'd often find myself maintaining my own version of something just to use a specific version, e.g. https://github.com/a-h/dotfiles/commit/6363839dca6a226bf3d064a4f70b355d90d7d48d or I'd have to pin a specific commit like the left hand side of this: https://github.com/a-h/dotfiles/commit/ea06709d306e80af920e02b359c5a95827d6e2f4

# How flakes help and how tool makers can help

Some tools are maintaining their own flake.nix files, so that you can use them as overlays. For example the xc tool maintains a flake here: https://github.com/joerdav/xc/blob/main/flake.nix

If I want to use it, I can run `nix shell github:joerdav/xc` and it will pull the latest commit of the repo, execute the build if required and I've got it.

But... I can also update my config to use @joerdav's flake as an overlay: https://github.com/a-h/dotfiles/commit/89ddf6254e911edd1030836d0d29dc534afa5e2a

The `flake.nix` takes in the stable nixos packages, but then I'm overlaying `xc` with the flake. Since @joerdav maintains the flake, I can remove my self-maintained version. Pinning is taken care of by the flake lock file, and because I can specific a specific git tag in the `github:joerdav/xc` input.

```nix
{
  inputs = {
    nixpkgs.url = "nixpkgs/nixos-22.11";
    home-manager = {
      url = "github:nix-community/home-manager";
      inputs.nixpkgs.follows = "nixpkgs";
    };
    xc = {
      url = "github:joerdav/xc";
      inputs.nixpkgs.follows = "nixpkgs";
    };
    darwin = {
      url = "github:lnl7/nix-darwin";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };

  outputs = inputs@{ self, nixpkgs, darwin, home-manager, xc, ... }:
    let
      pkgs = import nixpkgs {
        overlays = [
          (self: super: {
            xc = xc.packages.aarch64-darwin.xc;
          })
        ];
      };
    in rec {
      homeConfigurations = {
        adrian-linux = home-manager.lib.homeManagerConfiguration {
          system = "x86_64-linux";
          inherit pkgs;
          modules = [ ./.config/nixpkgs/home.nix ];
        };
      };
      darwinConfigurations = {
        adrian-mac = darwin.lib.darwinSystem {
          system = "aarch64-darwin";
          specialArgs = {inherit inputs;};
          inherit pkgs;
          modules = [ ./.config/nixpkgs/darwin-configuration.nix ];
        };
      };
    };
}
```

Since I wanted to use the latest `goreleaser` today, I created a flake for it at https://github.com/a-h/nix-goreleaser for personal use.

If I run `nix shell github:a-h/nix-goreleaser` I get a shell with the latest commit.

If I run `nix run github:a-h/nix-goreleaser`, it runs `goreleaser` from the latest commit of the flake.

Unlike with nixpkgs, I can tag versions, so I have tagged v1.15.2 and v1.16.0. Now I can run `nix shell github:a-h/nix-goreleaser/v1.15.2` to get that version and `nix shell github:a-h/nix-goreleaser/v1.16.0` to get the other.

I think this is pretty cool.

# Binary distribution vs source

There was some discussion in the other issues related to nix support around wanting to ship binaries. As it happens, I didn't bother to look at the existing nixpkgs configuration and jumped straight into making my own, and gave up when the tests failed (due to goreleaser tests expecting to be in a git repo), and didn't notice that the existing nixpkgs configuration dealt with it by setting `doCheck = false;`

So... I made a hacky binary distribution version that really isn't in the spirit of nix, but got me working this morning. FWIW, the existing nixpkgs config works great, and is more appealing than binary distribution to me, but if you want to go that route, my example shows it's possible:

```nix
{ pkgs, stdenv, fetchurl, autoPatchelfHook }:

let
  mapPlatformToName = platform:
    if platform.isDarwin then "Darwin"
    else if platform.isLinux then "Linux"
    else platform.parsed.kernel.name;

  suffixToHash = {
    # Use `print-hashes.sh` to generate the list below
    Linux_x86_64 = "19n5ywhg8gdav611w0iil1drnr4l1ba7akb8andi9fk54h8r70a9";
    Linux_arm64 = "0j28cddj5vvw6ni0hpgzazsn5qmg4pmmm23vqlrw535xvixi40jw";
    Darwin_x86_64="0mfsv7r2gm8zichc3rgfx45ivwfcvw6yjnl19rwmh0damir5c6yr";
    Darwin_arm64 = "1z6na7q4ws1mxbyiimp8vagb0wvfl9pqd4pvcisz62hw7wkza2fr";
  };

  mapPlatformToArchitecture = platform: {
    "x86_64" = "x86_64";
    "aarch64" = "arm64";
  }.${platform.parsed.cpu.name} or (throw "Unsupported CPU ${platform.parsed.cpu.name}");

  version = "v1.16.0";

  mapPlatformToSuffix = platform: "${mapPlatformToName platform}_${mapPlatformToArchitecture platform}";

  suffix = mapPlatformToSuffix stdenv.hostPlatform;
in
stdenv.mkDerivation {
  pname = "goreleaser";
  version = version;
  src = fetchurl {
    url = "https://github.com/goreleaser/goreleaser/releases/download/${version}/goreleaser_${suffix}.tar.gz";
    sha256 = suffixToHash.${suffix} or (throw "Missing hash for suffix ${suffix}");
  };
  sourceRoot = "."; # goreleaser doesn't have a folder in the tar.gz file.
  nativeBuildInputs = if pkgs.stdenv.isLinux
    then [ autoPatchelfHook ]
    else [];
  installPhase = ./install.sh;
  system = builtins.currentSystem;
}
```

# My thoughts

I currently think it would be best to maintain a `flake.nix` file in the `goreleaser` project, much like the https://github.com/joerdav/xc project does. Flakes seem to be the direction of travel, and it will keep the nix ecosystem up to date immediately when a new goreleaser release hits.

**Comment by caarlos0** [comment]
Thanks for the writeup, @a-h!

I think the main advantages to being able to install binaries are being able to install closed source software, and having the build flags/etc in a single place (in our particular case, `.goreleaser.yml`).

So, that's why I initially wanted to go that way... we might make it optional to our users, though.

That being said, to make this work, goreleaser would basically need to create a nix file in a given github repository, and that's it? Is there any specific filenames to be aware of?

It might be another pipe similar to brew, krew, etc...

**Comment by a-h** [comment]
No problem!

Yes, I think the job is to maintain a `flake.nix` in a repo using a pipeline, and tag it with releases. That's what `xc` does.

It uses a Go template to populate the `xc.nix` file that's then referenced by the `flake.nix`: 

* Template to update the `xc.nix`:
  * https://github.com/joerdav/xc/blob/main/xc.nix.tmpl
* Script to update the `xc.nix`
  * https://github.com/joerdav/xc/blob/main/update-nix.sh
* Flake that uses the `xc.nix` defition
  * https://github.com/joerdav/xc/blob/main/flake.nix

**Comment by caarlos0** [comment]
awesome, so, I guess the next step is to define our template for it... and for that I'll need to learn some stuff :D 

BTW: if you have a path suggestion to learn all this let me know, otherwise I'll just jump into the documentation hehe

**Comment by a-h** [comment]
Yes, I think that's the way.

I don't think it's very easy to get into Nix, but I never tried to really sit down and learn it methodically. Like, I didn't really try to understand the language properly by building up my understanding of the language alone. I was trying to read examples, and not really understanding what the code was doing for a while. I tend to use IDE autocomplete to discover fields and capabilities, and Nix doesn't really lend itself to that, you've got to use the repl and read the source code to find out what's available.

I use it for my systems, but it's just a tool for me, I don't want to spend too much time on it, I want to build other things.

I think it would be worth getting feedback on my proposal of a way forward from the the listed maintainers of the `goreleaser` Nix package. That's:

@c0deaddict
@endocrimes
@sarcasticadmin
@techknowlogick
@developer-guy
@caarlos0

I'd love to hear your thoughts on this proposal.

**Comment by caarlos0** [comment]
fwiw, spent most of this weekend playing with nix, home-manager, etc... 

Does anyone have any thoughts on distributing from the pre-compiled binaries as @a-h provided?

**Comment by caarlos0** [comment]
okay, fwiw I experimented a bit with this and NUR today... so now you can install goreleaser pro on nix!

for instance, using flake and home manager:


```nix
{
  description = "Home Manager configuration of Carlos Becker";

  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";

    home-manager = {
      url = "github:nix-community/home-manager";
      inputs.nixpkgs.follows = "nixpkgs";
    };

    mkAlias = {
      url = "github:cdmistman/mkAlias";
      inputs.nixpkgs.follows = "nixpkgs";
    };

    nix-index-database = {
      url = "github:Mic92/nix-index-database";
      inputs.nixpkgs.follows = "nixpkgs";
    };

    nur.url = "github:nix-community/NUR";
    caarlos0-nur.url = "github:caarlos0/nur";
  };

  outputs = { nur, caarlos0-nur, home-manager, nix-index-database, ... }@inputs:
    let
      overlays = final: prev: {
        nur = import nur {
          nurpkgs = prev;
          pkgs = prev;
          repoOverrides = { caarlos0 = import caarlos0-nur { pkgs = prev; }; };
        };
        # ... your other overlays
      };
    in {
        "user@host" = home-manager.lib.homeManagerConfiguration {
          pkgs = inputs.nixpkgs.legacyPackages.x86_64-linux;
          extraSpecialArgs = { inherit inputs; };
          modules = [
            ({ config, ... }: {
              config = { nixpkgs.overlays = [ overlays ]; };
            })
            ({ pkgs, ... }: {
              home.packages = [ pkgs.nur.repos.caarlos0.goreleaser-pro ];
            })
          ];
        };
      };
    };
}
```

The source is heavily based on @a-h examples, if you wanna check it out: https://github.com/caarlos0/nur

If you think this is a good approach, we could something that does this automatically into goreleaser...

User would still have to set up their NUR repo, and add their packages to the `default.nix` file... but wouldn't need to change anything on updates.

Another thing to investigate is if we can/its worth it reverse-engineer `nix-prefetch-url`... if not, we can assume the user will have to have it installed.


Thoughts?

**Comment by a-h** [comment]
I do think it makes sense for the Pro version to be distributed as a flake that installs binary files.

I don't know whether nixpkgs would take a binary distribution, probably worth giving that a try. I hadn't heard about NUR, but it sounds a bit like PPAs in the Ubuntu world, and it seems like a good way to make the Pro flake discoverable (other than goreleaser's own docs).

**Comment by caarlos0** [comment]
we could also support PRing things like this: https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/archivers/rar/default.nix

**Comment by blaggacao** [comment]
```nix
{
  description = "{{ .ProjectDescription }}";
  outputs = { ... }: {
    overlays = rec {
      default = {{ .Name }};
      {{ .Name }} = prev: final: {
        {{ .Name }} = prev.buildGo{Package,Application} {
          pname = "{{ .Name }}";
          version = "{{ .Version }}";
          vendorHash = "{{ .VendorHash }}";
          src = ./.;
          /* more, e.g. ldflags, etc */
        };
      };
    };
  };
}

**Comment by rvolosatovs** [comment]
> ```nix
> {
>   description = "{{ .ProjectDescription }}";
>   outputs = { ... }: {
>     overlays = rec {
>       default = {{ .Name }};
>       {{ .Name }} = final: prev: {
>         {{ .Name }} = prev.buildGo{Package,Application} {
>           pname = "{{ .Name }}";
>           version = "{{ .Version }}";
>           vendorHash = "{{ .VendorHash }}";
>           src = ./.;
>           /* more, e.g. ldflags, etc */
>         };
>       };
>     };
>   };
> }
> ```

While this sounds great on paper, I'd much rather see a `flake.nix` in this repository exporting a `lib`, which could parse the goreleaser manifest and provide a set of packages (or an overlay, or both). Here's an example in that direction within https://github.com/bytecodealliance/wit-deps, which is a Wasm WIT dependency manager written in Rust: https://github.com/bytecodealliance/wit-deps/blob/b4314067818a45413fc6c366983f19a80b2626fa/flake.nix#L32-L136

(shameless plug ahead) I am author and maintainer of https://github.com/rvolosatovs/nixify, which essentially generates Nix derivations from project source code, such as static binary builds and OCI images containing them. It currently only supports Rust and relies on `Cargo.toml`.

I am planning to add Go integration and doing so via GoReleaser manifest as an option would be very interesting for me.

So my proposal would be something like this:

`github:goreleaser/goreleaser` flake (i.e. `flake.nix` file in the root of this repository) exposing something along the lines of:
```nix
{
   lib.mkPackages = {
      src, # path to source containing `.goreleaser.yaml`
      pkgs, # nixpkgs package set
   }: {
      "${pname}-amd64-linux" = pkgs.stdenv.mkDerivation {
           # amd64 linux build if configured in `.goreleaser.yaml`
      };
      "${pname}-arm64-darwin" = pkgs.stdenv.mkDerivation {
           # arm64 mac build if configured in `.goreleaser.yaml`
      };
   }
}
```

This should require no changes to GoReleaser implementation itself, just the addition of the Nix library to the repository.

So then downstream consumers of GoReleaser could do something like this:
```nix
{
   inputs.goreleaser.url = github:goreleaser/goreleaser;
   outputs = {
      goreleaser,
      nixpkgs,
   }: {
      # this would contain everything from `build` in `./.goreleaser.yml` buildable by `x86_64-linux`
      packages.x86_64-linux = goreleaser.lib.mkPackages {
         src = ./.;
         pkgs = import nixpkgs {};
      };
   }
```

I'm happy to contribute to this or at least start with a simple MVP sketch

**Comment by blaggacao** [comment]
@rvolosatovs excellent approach / idea!

However, I see this as an added value on a lower layer.

Our user personas may be alomg those lines:

**Gopher :heart: Nixer**

A go person somewhat has heard of Nix and acknowledges its benefits and would like to offer their users `nix run github:my/repo -- ...` capabilities.

This user persona wants `goreleaser` to place a `flake.nix` at the repo and/or package root.

_In my above example, I started with an overlay approach as the foundational building block to avoid the cumulative prohibitive cost of yet-another-nixpkgs-evaluation. It only satifies this user persona with the addition of a `nixpkgs` pin._

**Nixer :heart: Gopher**

This persona sees a non-canonically nixified source and is not ok with that fact. She wants to plug in the software artifact into her Nix dependency layer.

She would enjoy a canonical transpiler of goreleaser instructions into _native_ nix instructions (IFD would be too costly).

This is the high level objective of the proposed `lib.mkPackages` utility function.

Currently native data ingestion is limited to the following data formats `builtins.{fromJSON,fromTOML,fromXML}`.

The evaluation cost of IFD (i.e. use whatever tool you like for data ingestion) is prohibitive as an implementation fundament.

There is a PR being considered by The Nix Team to include `builtins.fromYAML`, but if goreleaser needs other data sources (e.g. `go.sum), the only option for transpilation is code generation.

**Comment by caarlos0** [comment]
another thing we shouldn't forget: goreleaser is also used to build and release nonfree, closed source software, in which case nix should only download the pre-compiled binaries and place them in the right places.

Since that usecase would also work for OSS software, my first idea would be to make goreleaser output a `.nix` file that would do just that, probably somewhat templateable as the homebrew and other integrations in goreleaser are.

That file could then be pushed to a NUR and/or PRed into nixpkgs.

Now, onto the build: basically the options currently available in `buildGoModule` would be sufficient (as far as I know), so we could also generate flakes to compile from source, as long as there's a way to programmatically get the checksum of the vendor folder so we can template it properly.

That said, I'm not sure writing a whole transpiler makes sense, as interesting as it sounds 🤔 

--- 

## summarizing 

IMHO, 2 possible ways, maybe we can even implemente both:

- create a `flake.nix` like [this](https://github.com/caarlos0/nur/blob/master/pkgs/xdg-open-svc/default.nix) from `.goreleaser.yml` in the root of the repository using `buildGoModule`
  - this won't work with nonfree software
  - this can be PRed/pushed into a NUR and/or nixpkgs
- create a `pkg.nix` that downloads and installs the pre-compiled binaries like [this](https://github.com/caarlos0/nur/blob/master/pkgs/goreleaser-pro/default.nix)
  - works on free and nonfree projects
  - can be pushed/pr to a NUR
    - not sure if we should make it too easy to PR to nixpkgs too

Seems like simple enough approaches, unless I'm missing something.

thoughts?

**Comment by blaggacao** [comment]
Re private packages, there is a nix builtin for that use case, probably somewhat disguised. Come back in 30 mins and I'll have probably found the reference... : 😊

**Comment by caarlos0** [comment]
@blaggacao not necessarily speaking of private packages, as shipping binaries with private source code, e.g. how rar is distributed in nixpkgs - the pkg is public, the source to rar isn't (afaik)

**Comment by roberth** [comment]
**Flakes**

These are still experimental, but already embraced by the community. While the feature doesn't change rapidly, it does change.

**vendorHash**

This is a stop-gap solution. Nix allows arbitrarily complex network access in parts of the build graph that produce an output with a "hardcoded" hash. This should be used for fairly reliable processes, such as fetching a file, or a git commit's contents. `go mod vendor` is close to repeatable, but subtle changes to the vendoring logic will require a change in `vendorHash`. This means that in some cases, the hashes may not be compatible with a newer Go.
We have plans that may remove the need for `vendorHash`, but I can make no promises. (for the nix nerds: "dynamic derivations"/rfc 92, to produce many smaller fetchers without IFD)

**overlays**

Overlays were originally meant for making modifications to existing packages; modifications that propagate through the entire nixpkgs set, including dependencies of other packages.

Some people will use them because they provide a convenient place for custom packages. This was especially relevant before flakes, when a pattern emerged to reuse `pkgs` for everything.
However, this has led to surprises. If all you need is to use an updated or patched package for one purpose, you shouldn't have to use it everywhere, and you shouldn't have to deal with build errors or unnecessary rebuilds.

In other words, overlays, not unlike `/usr` can be too powerful. That's not to say that they're not a valid solution, but it'd be good to have an alternative.

Such an alternative is the flake `packages` attribute. These aren't combined into any other structures unless you tell Nix to.
By default they use a known good nixpkgs version for dependencies, and users of the flake can pass their own version of nixpkgs into the flake if they desire (fewer nixpkgs versions -> fewer distinct versions -> fewer distinct store paths).


**Expectations**

Nix and the expressions written in it are still evolving. It is a certainty that any nix integration will have to change with it, but as you can see there's plenty of interest.

**Comment by caarlos0** [comment]
started working on something here: https://github.com/goreleaser/goreleaser/pull/4012
